### PR TITLE
add_benchmark_mode

### PR DESCRIPTION
**Pull request by mdrozdowski1996**

### CODE REVIEW

File: tftrt/examples/image-classification/image_classification.py [hunk 1]
```diff
@@ -30,6 +30,10 @@
 import official.resnet.imagenet_main
 from preprocessing import inception_preprocessing, vgg_preprocessing
 
+tf.enable_eager_execution()
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: why do we have eager execution here?

File: tftrt/examples/image-classification/image_classification.py [hunk 2]
```diff
@@ -40,7 +44,7 @@ def __init__(self, batch_size, num_records, display_every):
 
     def begin(self):
         self.start_time = time.time()
-
+
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Please remove this whitespace.

File: tftrt/examples/image-classification/image_classification.py [hunk 3]
```diff
@@ -562,6 +615,8 @@ def get_frozen_graph(
         help='workspace size in bytes')
     parser.add_argument('--cache', action='store_true',
         help='If set, graphs will be saved to disk after conversion. If a converted graph is present on disk, it will be loaded instead of building the graph again.')
+    parser.add_argument('--mode', type=str, default='validation',
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Please add options for this so that argparse automatically checks if the users gives a wrong value.

File: tftrt/examples/image-classification/image_classification.py (hunk 1 above)
Comment (mdrozdowski1996) [reply tftrt/examples/image-classification/image_classification.py]: You are right, we don't need to use it. I will change that.

File: tftrt/examples/image-classification/image_classification.py [hunk 4]
```diff
@@ -106,34 +127,60 @@ def eval_input_fn():
...
     estimator = tf.estimator.Estimator(
         model_fn=model_fn,
         config=tf.estimator.RunConfig(session_config=tf_config),
         model_dir='model_dir')
-    results = estimator.evaluate(eval_input_fn, steps=num_iterations, hooks=[logger])
-    
-    # Gather additional results
+    if mode == 'validation':
+        print("evaluate")
+        results = estimator.evaluate(input_fn, steps=num_iterations, hooks=[logger])
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: results -> validation_results

File: tftrt/examples/image-classification/image_classification.py [hunk 5]
```diff
@@ -106,34 +127,60 @@ def eval_input_fn():
...
+    else:
+        print("predict")
+        results = estimator.predict(input_fn, predict_keys=["logits", "classes"],  hooks=[logger])
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: results -> generator or something like that

File: tftrt/examples/image-classification/image_classification.py [hunk 6]
```diff
@@ -519,6 +566,12 @@ def get_frozen_graph(
 
     return frozen_graph, num_nodes, times, graph_sizes
 
+def load_and_preprocess_image(path):
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Remove this.

File: tftrt/examples/image-classification/image_classification.py [hunk 7]
```diff
@@ -83,6 +95,15 @@ def model_fn(features, labels, mode):
     # preprocess function for input data
     preprocess_fn = get_preprocess_fn(model)
 
+    def load_and_preprocess_image(path):
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Can we combine `load_and_preprocess_image` and `get_preprocess_fn`?
It would need a mode argument.

File: tftrt/examples/image-classification/image_classification.py [hunk 8]
```diff
@@ -53,7 +57,8 @@ def after_run(self, run_context, run_values):
                 self.batch_size / self.iter_times[-1]))
 
 def run(frozen_graph, model, data_files, batch_size,
-    num_iterations, num_warmup_iterations, use_synthetic=False, display_every=100, run_calibration=False):
+    num_iterations, num_warmup_iterations, use_synthetic, display_every=100, run_calibration=False, mode='validation'):
+
     """Evaluates a frozen graph
     
     This function evaluates a graph on the ImageNet validation set.
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Write docs for all the args.

File: tftrt/examples/image-classification/image_classification.py [hunk 9]
```diff
@@ -72,9 +77,16 @@ def model_fn(features, labels, mode):
...
             return_elements=['logits:0', 'classes:0'],
             name='')
-        loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits_out)
-        accuracy = tf.metrics.accuracy(labels=labels, predictions=classes_out, name='acc_op')
+        if mode == tf.estimator.ModeKeys.PREDICT:
+            predictions = {}
+            with tf.name_scope('Predictions'):
+                predictions['logits'] = logits_out
+                predictions['classes'] = classes_out
+            print(predictions)
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Don't print.

File: tftrt/examples/image-classification/image_classification.py [hunk 10]
```diff
@@ -72,9 +77,16 @@ def model_fn(features, labels, mode):
             input_map={'input': features},
             return_elements=['logits:0', 'classes:0'],
             name='')
-        loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits_out)
-        accuracy = tf.metrics.accuracy(labels=labels, predictions=classes_out, name='acc_op')
+        if mode == tf.estimator.ModeKeys.PREDICT:
+            predictions = {}
+            with tf.name_scope('Predictions'):
+                predictions['logits'] = logits_out
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: I think we can also inline these: 

`tf.estimator.EstimatorSpec(mode=mode, predictions={'logits':logits_out, 'classes':classes_out})`

**Comment by pooyadavoodi** [comment]
Please post an example of the log here after all the comments are resolved.

File: tftrt/examples/image-classification/image_classification.py (hunk 10 above)
Comment (trevor-m) [reply tftrt/examples/image-classification/image_classification.py]: It might be better to only fetch classes to reduce IO.

Comment (pooyadavoodi) [reply tftrt/examples/image-classification/image_classification.py]: oh yes, that's a good idea.

File: tftrt/examples/image-classification/image_classification.py (hunk 3 above)
Comment (mdrozdowski1996) [reply tftrt/examples/image-classification/image_classification.py]: What should be the valid options?

**Comment by mdrozdowski1996** [comment]
[output_tftrt_fp32_bs8_inception_v3_dynamic_op=False.txt](https://github.com/tensorflow/tensorrt/files/2838501/output_tftrt_fp32_bs8_inception_v3_dynamic_op.False.txt)

File: tftrt/examples/image-classification/image_classification.py [hunk 11]
```diff
@@ -30,9 +30,12 @@
 import official.resnet.imagenet_main
 from preprocessing import inception_preprocessing, vgg_preprocessing
 
+
+
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Remove an extra newline.

File: tftrt/examples/image-classification/image_classification.py [hunk 12]
```diff
@@ -51,9 +54,13 @@ def after_run(self, run_context, run_values):
             print("    step %d/%d, iter_time(ms)=%.4f, images/sec=%d" % (
                 current_step, self.num_steps, duration * 1000,
                 self.batch_size / self.iter_times[-1]))
+        self.iteration_number += 1
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Isn't it better to make a new Hook for `iteration_number`?

We made `LoggerHook` to improve logging, but `iteration_number` is actually used for stopping the benchmark.

**Comment by pooyadavoodi** [comment]
Please resolve merge conflicts.
We just added another Hook that's used for benchmarking. It's called DurationHook.

Some parts of your PR would need to use DurationHook.

File: tftrt/examples/image-classification/image_classification.py [hunk 13]
```diff
@@ -65,23 +72,32 @@ def run(frozen_graph, model, data_files, batch_size,
     data_files: List of TFRecord files used for inference
     batch_size: int, batch size for TensorRT optimizations
     num_iterations: int, number of iterations(batches) to run for
+    num_warmup_iterations: int, number of iteration(batches) to exclude from benchmark measurments
+    use_synthetic: bool, if true run using real data, otherwise synthetic
+    display_every: int, print log every @display_every iteration
+    run_calibration: bool, run using calibration or not (only int8 precision)
+	mode: validation - using estimator.evaluate with accuracy measurments,
+          benchmark - using estimator.predict
     """
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Please add a check after this line to ensure `mode` has one of the two values: `validation` or `benchmark`. Otherwise raise an exception.

File: tftrt/examples/image-classification/image_classification.py [hunk 14]
```diff
@@ -106,28 +122,51 @@ def eval_input_fn():
...
         model_fn=model_fn,
         config=tf.estimator.RunConfig(session_config=tf_config),
         model_dir='model_dir')
-    results = estimator.evaluate(eval_input_fn, steps=num_iterations, hooks=[logger])
-    
-    # Gather additional results
+    if mode == 'validation':
+        results = estimator.evaluate(input_fn, steps=num_iterations, hooks=[logger])
+    else:
+        results = {}
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Please move `results = {}` to outside of the if/else block before `if mode == 'validation':`.

File: tftrt/examples/image-classification/image_classification.py [hunk 15]
```diff
@@ -106,28 +122,51 @@ def eval_input_fn():
...
+        prediction_results = estimator.predict(input_fn, predict_keys=["classes"],  hooks=[logger])
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: We would need to use `DurationHook` in `estimator.predict` (but not in estimator.evaluate)`.

File: tftrt/examples/image-classification/image_classification.py [hunk 16]
```diff
@@ -106,28 +122,51 @@ def eval_input_fn():
...
+        x = 0
+        for i in prediction_results:
+            x += 1
+    #: Gather additional results
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Please remove the extra `:` from the comment.

File: tftrt/examples/image-classification/image_classification.py [hunk 17]
```diff
@@ -183,7 +222,7 @@ def get_netdef(model):
     """
     Creates the dictionary NETS with model names as keys and NetDef as values.
     Returns the NetDef corresponding to the model specified in the parameter.
-
+3
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Please remove this. I don't think it's intended?

File: tftrt/examples/image-classification/image_classification.py (hunk 13 above)
Comment (pooyadavoodi) [reply tftrt/examples/image-classification/image_classification.py]: Similarly in other functions that use `mode` throughout the script.

File: tftrt/examples/image-classification/image_classification.py (hunk 3 above)
Comment (pooyadavoodi) [reply tftrt/examples/image-classification/image_classification.py]: I would define two options for now: `validation` and `benchmark`.

File: tftrt/examples/image-classification/image_classification.py [hunk 18]
```diff
@@ -574,20 +628,32 @@ def get_frozen_graph(
...
+        raise ValueError('invalid mode argument (must be validation or benchmark)')
+    
     def get_files(data_dir, filename_pattern):
         if data_dir == None:
             return []
         files = tf.gfile.Glob(os.path.join(data_dir, filename_pattern))
         if files == []:
-            raise ValueError('Can not find any files in {} with pattern "{}"'.format(
-                data_dir, filename_pattern))
+            raise ValueError('Can not find any files in {} with pattern "{}"'.format(data_dir, filename_pattern))
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Why is this line changed?

File: tftrt/examples/image-classification/image_classification.py [hunk 19]
```diff
@@ -574,20 +628,32 @@ def get_frozen_graph(
...
         return files
 
-    validation_files = get_files(args.data_dir, 'validation*')
+    if args.mode == "validation":
+        validation_files = get_files(args.data_dir, 'validation*')
+    else:    
+        val_paths = get_files(args.data_dir, "n*")
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Is it necessary to have the filenames start with `n*`?
Why can't we instead read every file in the directory?

Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Please rename `vali_paths` and validation_files` in the `else` block because they are for benchmark mode.

File: tftrt/examples/image-classification/image_classification.py [hunk 20]
```diff
@@ -52,29 +52,34 @@ def after_run(self, run_context, run_values):
                 current_step, self.num_steps, duration * 1000,
                 self.batch_size / self.iter_times[-1]))
 
-class DurationHook(tf.train.SessionRunHook):
+class BenchmarkHook(tf.train.SessionRunHook):
     """Limits run duration"""
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Could you update the docs here explaining what `target_duration` and `iteration_limit` are exactly.

File: tftrt/examples/image-classification/image_classification.py [hunk 21]
```diff
@@ -52,29 +52,34 @@ def after_run(self, run_context, run_values):
                 current_step, self.num_steps, duration * 1000,
                 self.batch_size / self.iter_times[-1]))
 
-class DurationHook(tf.train.SessionRunHook):
+class BenchmarkHook(tf.train.SessionRunHook):
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Could you update the docs here explaining what `target_duration` and `iteration_limit` are exactly.

File: tftrt/examples/image-classification/image_classification.py [hunk 22]
```diff
@@ -52,29 +52,34 @@ def after_run(self, run_context, run_values):
...
     """Limits run duration"""
-    def __init__(self, target_duration):
+    def __init__(self, target_duration=None, iteration_limit=None):
         self.target_duration = target_duration
         self.start_time = None
+        self.iteration_number = 0
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Could you please rename `iteration_number` to `current_iteration`?

Comment (mdrozdowski1996) [reply tftrt/examples/image-classification/image_classification.py]: sure

File: tftrt/examples/image-classification/image_classification.py [hunk 23]
```diff
@@ -52,29 +52,34 @@ def after_run(self, run_context, run_values):
...
+
+        if self.iteration_limit:
+            self.iteration_number += 1
+            if self.iteration_number >= self.iteration_limit:
+                run_context.request_stop()
 
 def run(frozen_graph, model, data_files, batch_size,
-    num_iterations, num_warmup_iterations, use_synthetic, display_every=100, run_calibration=False,
-    target_duration=None):
+    num_iterations, num_warmup_iterations, use_synthetic, display_every=100, run_calibration=False,
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Remove trailing whitespace.

File: tftrt/examples/image-classification/image_classification.py [hunk 24]
```diff
@@ -86,23 +91,32 @@ def run(frozen_graph, model, data_files, batch_size,
     data_files: List of TFRecord files used for inference
     batch_size: int, batch size for TensorRT optimizations
     num_iterations: int, number of iterations(batches) to run for
+    num_warmup_iterations: int, number of iteration(batches) to exclude from benchmark measurments
+    use_synthetic: bool, if true run using real data, otherwise synthetic
+    display_every: int, print log every @display_every iteration
+    run_calibration: bool, run using calibration or not (only int8 precision)
+	mode: validation - using estimator.evaluate with accuracy measurments,
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Remove extra indentation before `mode`.

File: tftrt/examples/image-classification/image_classification.py [hunk 25]
```diff
@@ -127,28 +141,57 @@ def eval_input_fn():
...
-    duration_hook = DurationHook(target_duration)
-    results = estimator.evaluate(eval_input_fn, steps=num_iterations, hooks=[logger, duration_hook])
-
+    results = {}
+    if mode == 'validation':
+        results = estimator.evaluate(input_fn, steps=num_iterations, hooks=[logger])
+    elif mode == 'benchmark':
+        benchmark_hook = BenchmarkHook(target_duration, (num_records + batch_size - 1)/batch_size)
+        prediction_results = estimator.predict(input_fn, predict_keys=["classes"],  hooks=[logger, benchmark_hook])
+        x = 0
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Would something like this work?

`prediction_results = [p for p in estimator.predict(input_fn, predict_keys=["classes"],  hooks=[logger, benchmark_hook])]`

File: tftrt/examples/image-classification/image_classification.py [hunk 26]
```diff
@@ -127,28 +141,57 @@ def eval_input_fn():
...
         config=tf.estimator.RunConfig(session_config=tf_config),
         model_dir='model_dir')
-    duration_hook = DurationHook(target_duration)
-    results = estimator.evaluate(eval_input_fn, steps=num_iterations, hooks=[logger, duration_hook])
-
+    results = {}
+    if mode == 'validation':
+        results = estimator.evaluate(input_fn, steps=num_iterations, hooks=[logger])
+    elif mode == 'benchmark':
+        benchmark_hook = BenchmarkHook(target_duration, (num_records + batch_size - 1)/batch_size)
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Shouldn't we use `num_iterations` for the limit?

File: tftrt/examples/image-classification/image_classification.py (hunk 23 above)
Comment (mdrozdowski1996) [reply tftrt/examples/image-classification/image_classification.py]: But this is the whitespace after comma.

File: tftrt/examples/image-classification/image_classification.py [hunk 27]
```diff
@@ -127,28 +141,57 @@ def eval_input_fn():
...
-    duration_hook = DurationHook(target_duration)
-    results = estimator.evaluate(eval_input_fn, steps=num_iterations, hooks=[logger, duration_hook])
-
+    results = {}
+    if mode == 'validation':
+        results = estimator.evaluate(input_fn, steps=num_iterations, hooks=[logger])
+    elif mode == 'benchmark':
+        benchmark_hook = BenchmarkHook(target_duration, num_iterations)
+        prediction_results = [p for p in estimator.predict(input_fn, predict_keys=["classes"],  hooks=[logger, benchmark_hook])]
+        #x = 0
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Remove comments.

File: tftrt/examples/image-classification/image_classification.py [hunk 28]
```diff
@@ -127,28 +141,57 @@ def eval_input_fn():
...
         config=tf.estimator.RunConfig(session_config=tf_config),
         model_dir='model_dir')
-    duration_hook = DurationHook(target_duration)
-    results = estimator.evaluate(eval_input_fn, steps=num_iterations, hooks=[logger, duration_hook])
-
+    results = {}
+    if mode == 'validation':
+        results = estimator.evaluate(input_fn, steps=num_iterations, hooks=[logger])
+    elif mode == 'benchmark':
+        benchmark_hook = BenchmarkHook(target_duration, num_iterations)
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Could you change it to this:
`BenchmarkHook(target_duration=target_duration, iteration_limit=num_iterations)`

File: tftrt/examples/image-classification/image_classification.py [hunk 29]
```diff
@@ -127,28 +141,57 @@ def eval_input_fn():
...
+        prediction_results = [p for p in estimator.predict(input_fn, predict_keys=["classes"],  hooks=[logger, benchmark_hook])]
+        #x = 0
+        #for i in prediction_results:
+        #    x += 1
+    else:
+        raise ValueError("Mode must be equal 'validation' or 'benchmark'")
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: "Mode must be equal 'validation' or 'benchmark'" --> "mode must be either 'validation' or 'benchmark'"

File: tftrt/examples/image-classification/image_classification.py (hunk 23 above)
Comment (pooyadavoodi) [reply tftrt/examples/image-classification/image_classification.py]: Right, but it's at the end of the line. pylint should fix this.

File: tftrt/examples/image-classification/image_classification.py [hunk 30]
```diff
@@ -292,14 +333,14 @@ def deserialize_image_record(record):
         text    = obj['image/class/text']
         return imgdata, label, bbox, text
 
-def get_preprocess_fn(model, mode='classification'):
+def get_preprocess_fn(model, mode='validation'):
     """Creates a function to parse and process a TFRecord using the model's parameters
 
     model: string, the model name (see NETS table)
     mode: string, whether the model is for classification or detection
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Please update the doc for `mode`.

File: tftrt/examples/image-classification/image_classification.py (hunk 3 above)
Comment (pooyadavoodi) [reply tftrt/examples/image-classification/image_classification.py]: This is not done yet.

File: tftrt/examples/image-classification/image_classification.py [hunk 31]
```diff
@@ -598,20 +656,33 @@ def get_frozen_graph(
     if args.num_calib_inputs < args.batch_size:
         raise ValueError('--num_calib_inputs must not be smaller than --batch_size'
             '({} <= {})'.format(args.num_calib_inputs, args.batch_size))
-
+    if args.mode == 'validation' and args.use_synthetic:
+        raise ValueError('Cannot use both validation mode and synthetic dataset')
+    if args.mode != 'validation' and args.mode != 'benchmark':
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Once you add the options to `add_argument` then we don't need this here.

File: tftrt/examples/image-classification/image_classification.py [hunk 32]
```diff
@@ -598,20 +656,33 @@ def get_frozen_graph(
...
+        raise ValueError('invalid mode argument (must be validation or benchmark)')
+    
     def get_files(data_dir, filename_pattern):
         if data_dir == None:
             return []
         files = tf.gfile.Glob(os.path.join(data_dir, filename_pattern))
         if files == []:
             raise ValueError('Can not find any files in {} with pattern "{}"'.format(
-                data_dir, filename_pattern))
+			    data_dir, filename_pattern))
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Remove tabs here.

File: tftrt/examples/image-classification/image_classification.py [hunk 33]
```diff
@@ -598,20 +656,33 @@ def get_frozen_graph(
...
 
-    validation_files = get_files(args.data_dir, 'validation*')
+    if args.mode == "validation":
+        data_files = get_files(args.data_dir, 'validation*')
+    elif args.mode == "benchmark":    
+        benchmark_paths = get_files(args.data_dir, "*")
+        data_files = []
+        for path in benchmark_paths:
+            data_files = data_files + get_files(path, "*")
+        if args.num_iterations != None:
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: I don't think this is needed here. Later we only run inference the right number of times.

File: tftrt/examples/image-classification/image_classification.py [hunk 34]
```diff
@@ -598,20 +656,33 @@ def get_frozen_graph(
...
             raise ValueError('Can not find any files in {} with pattern "{}"'.format(
-                data_dir, filename_pattern))
+			    data_dir, filename_pattern))
         return files
 
-    validation_files = get_files(args.data_dir, 'validation*')
+    if args.mode == "validation":
+        data_files = get_files(args.data_dir, 'validation*')
+    elif args.mode == "benchmark":    
+        benchmark_paths = get_files(args.data_dir, "*")
```
Comment (pooyadavoodi) [inline tftrt/examples/image-classification/image_classification.py]: Something like this should work:

`data_files = [os.path.join(path, name) for path, _, files in os.walk(args.data_dir) for name in files]`

File: tftrt/examples/image-classification/image_classification.py (hunk 23 above)
Comment (pooyadavoodi) [reply tftrt/examples/image-classification/image_classification.py]: This is not fixed yet.

File: tftrt/examples/image-classification/image_classification.py (hunk 29 above)
Comment (pooyadavoodi) [reply tftrt/examples/image-classification/image_classification.py]: This is not fixed yet.

File: tftrt/examples/image-classification/image_classification.py (hunk 30 above)
Comment (pooyadavoodi) [reply tftrt/examples/image-classification/image_classification.py]: This is not fixed yet.

File: tftrt/examples/image-classification/image_classification.py (hunk 29 above)
Comment (mdrozdowski1996) [reply tftrt/examples/image-classification/image_classification.py]: ?
